Fix handling of APIserver errors when saving provisioned PVs.

jsafrane opened the pull request:
When API server crashes *after* saving a provisioned PV and before sending
200 OK, the controller tries to save the PV again. In this case, it gets
AlreadyExists error, which should be interpreted as success and not as error.

Especially, a volume that corresponds to the PV should not be deleted in the
underlying storage.

Fixes #44372

```release-note
NONE
```

@kubernetes/sig-storage-pr-reviews

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44719)
<!-- Reviewable:end -->

jsafrane commented:
@k8s-bot cvm gce e2e test this

justinsb commented on pkg/controller/volume/persistentvolume/pv_controller.go:
```diff
@@ -1360,14 +1361,19 @@ func (ctrl *PersistentVolumeController) provisionClaimOperation(claimObj interfa
 		glog.V(4).Infof("provisionClaimOperation [%s]: trying to save volume %s", claimToClaimKey(claim), volume.Name)
 		var newVol *v1.PersistentVolume
-		if newVol, err = ctrl.kubeClient.Core().PersistentVolumes().Create(volume); err == nil {
+		if newVol, err = ctrl.kubeClient.Core().PersistentVolumes().Create(volume); err == nil || apierrs.IsAlreadyExists(err) {
```
Nit: I'm wondering if this would read better if we didn't use the "inline" if statement, but rather just handled the 3 cases explicitly

justinsb commented on pkg/controller/volume/persistentvolume/pv_controller.go:
```diff
@@ -1360,14 +1361,19 @@ func (ctrl *PersistentVolumeController) provisionClaimOperation(claimObj interfa
 			// Save succeeded.
-			glog.V(3).Infof("volume %q for claim %q saved", volume.Name, claimToClaimKey(claim))
+			if err != nil {
+				glog.V(3).Infof("volume %q for claim %q already exists, reusing", volume.Name, claimToClaimKey(claim))
```
Review note: the name includes the UID of the PVC, so there is no (real) risk of a naming collision here.  :+1:

justinsb reviewed (commented):
This LGTM, but I think it needs lgtm.  saad is our most likely candidate IMO.

justinsb commented:
/assign @saad-ali 

/unassign @rootfs 
/unassign @justinsb

jsafrane commented:
rebased. @saad-ali, PTAL

jsafrane commented:
/retest

jsafrane commented:
this PR is still valid
/assign @rootfs @wongma7

wongma7 commented:
/lgtm

wongma7 commented:
needs hack/update-bazel.sh, 

W0828 18:12:52.853] 2017/08/28 18:12:52 missing strict dependencies:
W0828 18:12:52.853] 	pkg/controller/volume/persistentvolume/provision_test.go: import of k8s.io/kubernetes/pkg/api, which is not a direct dependency

jsafrane committed:
Fix handling of APIserver errors when saving provisioned PVs.

When API server crashes *after* saving a provisioned PV and before sending
200 OK, the controller tries to save the PV again. In this case, it gets
AlreadyExists error, which should be interpreted as success and not as error.

Especially, a volume that corresponds to the PV should not be deleted in the
underlying storage.

jsafrane commented:
rebased (= fixed bazel buid)

jsafrane commented:
/retest
